tsconsensus/tsconsensus.go

```diff
@@ -0,0 +1,447 @@
+	// of a consensus
+	r, err := startRaft(shutdownCtx, ts, &fsm, c.self, auth, cfg)
+	if err != nil {
+		return nil, err
+	}
+	c.raft = r
+
+	c.bootstrap(auth.AllowedPeers())
+
+	if cfg.ServeDebugMonitor {
+		srv, err = serveMonitor(&c, ts, netip.AddrPortFrom(c.self.hostAddr, cfg.MonitorPort).String())
+		if err != nil {
+			return nil, err
+		}
+		c.monitorHttpServer = srv
+	}
+
+	return &c, nil
+}
+
+func startRaft(shutdownCtx context.Context, ts *tsnet.Server, fsm *raft.FSM, self selfRaftNode, auth *authorization, cfg Config) (*raft.Raft, error) {
+	cfg.Raft.LocalID = raft.ServerID(self.id)
+
+	// no persistence (for now?)
+	logStore := raft.NewInmemStore()
+	stableStore := raft.NewInmemStore()
+	snapshots := raft.NewInmemSnapshotStore()
+
+	// opens the listener on the raft port, raft will close it when it thinks it's appropriate
+	ln, err := ts.Listen("tcp", raftAddr(self.hostAddr, cfg))
+	if err != nil {
+		return nil, err
+	}
+
+	logger := hclog.New(&hclog.LoggerOptions{
+		Name:   "raft-net",
+		Output: cfg.Raft.LogOutput,
+		Level:  hclog.LevelFromString(cfg.Raft.LogLevel),
+	})
+
+	transport := raft.NewNetworkTransportWithLogger(StreamLayer{
+		s:           ts,
+		Listener:    ln,
+		auth:        auth,
+		shutdownCtx: shutdownCtx,
+	},
+		cfg.MaxConnPool,
+		cfg.ConnTimeout,
+		logger)
+
+	return raft.NewRaft(cfg.Raft, *fsm, logStore, stableStore, snapshots, transport)
+}
+
+// A Consensus is the consensus algorithm for a tsnet.Server
+// It wraps a raft.Raft instance and performs the peer discovery
+// and command execution on the leader.
+type Consensus struct {
+	raft              *raft.Raft
+	commandClient     *commandClient
+	self              selfRaftNode
+	config            Config
+	cmdHttpServer     *http.Server
+	monitorHttpServer *http.Server
+	shutdownCtxCancel context.CancelFunc
+}
+
+// bootstrap tries to join a raft cluster, or start one.
+//
+// We need to do the very first raft cluster configuration, but after that raft manages it.
+// bootstrap is called at start up, and we are not currently aware of what the cluster config might be,
+// our node may already be in it. Try to join the raft cluster of all the other nodes we know about, and
+// if unsuccessful, assume we are the first and start our own.
+//
+// It's possible for bootstrap to return an error, or start a errant breakaway cluster.
+//
+// We have a list of expected cluster members already from control (the members of the tailnet with the tag)
+// so we could do the initial configuration with all servers specified.
+// Choose to start with just this machine in the raft configuration instead, as:
+//   - We want to handle machines joining after start anyway.
+//   - Not all tagged nodes tailscale believes are active are necessarily actually responsive right now,
+//     so let each node opt in when able.
+func (c *Consensus) bootstrap(targets views.Slice[*ipnstate.PeerStatus]) error {
```

Comment:
some version of this should probably run continuously, and some path for split-brain recovery